### PR TITLE
Fix ads upgrade-related browser tests

### PR DESCRIPTION
**Pull request by emerick**
Resolves https://github.com/brave/brave-browser/issues/15335

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: test/data/rewards-data/migration/PreferencesForVersion12WithRewardsAndAdsDisabled [hunk 1]
```diff
@@ -13,7 +13,7 @@
   "brave": {
     "brave_ads": {
       "ads_per_hour": "2",
-      "enabled": true,
+      "enabled": false,
```
Comment (emerick) [inline test/data/rewards-data/migration/PreferencesForVersion12WithRewardsAndAdsDisabled:16]: @tmancey This particular test was still failing even after my fix. Given that it's called `WithRewardsAndAdsDisabled` I'm thinking this needs to be `false` as I've done here, but let me know if that's wrong.

File: components/brave_ads/browser/ads_service_browsertest.cc [hunk 2]
```diff
@@ -999,6 +1017,19 @@ const BraveAdsUpgradePathParamInfo kTests[] = {
...
+  ASSERT_NE(nullptr, test_info);
+  const std::string test_name = test_info->name();
+
+  // Wait for Brave Rewards to be initialized before proceeding with
+  // tests that rely on Rewards
+  if ((test_name.find("WithRewardsEnabled_") != std::string::npos ||
+       test_name.find("WithRewardsAndAdsEnabled_") != std::string::npos) &&
+      !rewards_service_->IsInitialized()) {
+    WaitForRewardsInitialization();
+  }
```
Comment (mariospr) [inline components/brave_ads/browser/ads_service_browsertest.cc:1032]: Ah! So this was the missing piece then? LGTM!

**Review by mariospr — APPROVED** [review]
LGTM

**Review by tmancey — APPROVED** [review]
LGTM++

File: components/brave_ads/browser/ads_service_browsertest.cc (hunk 2 above)
Comment (emerick) [reply components/brave_ads/browser/ads_service_browsertest.cc:1032]: Yup, I believe so. We'll see what CI says. I have to make a small adjustment for Linux; it didn't seem to like that I derived the observer from the main test class (a `DCHECK(!g_browser_process)` was failing in test setup). Will just go with the code you originally had!

**Comment by emerick** [comment]
Windows and Android CI passed. Linux CI passed with expected/unrelated test failures. Mac/iOS CI failed on init, so restarting for those platforms.

**Comment by emerick** [comment]
CI passed for iOS, but Mac aborted. Rerunning Mac.

**Comment by emerick** [comment]
CI passed on Mac; ready to merge.

**Pull request merged**